metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,244 @@
+            if (Double.compare(scalingFactor, 0) != 0) {
+                RescalingConsumer consumer = new RescalingConsumer(scalingFactor, newValues);
+                values.forEach(consumer);
+                // make sure the counter is in sync with the number of stored samples.
+                newCount = new AtomicLong(consumer.count);
+            } else {
+                newCount = new AtomicLong();
+            }
+            return new State(newTick, newCount, newValues);
+        }
+    }
+
+    private static final class RescalingConsumer implements BiConsumer<Double, WeightedSample> {
+        private final double scalingFactor;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+        private long count;
+
+        RescalingConsumer(double scalingFactor, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.scalingFactor = scalingFactor;
+            this.values = values;
+        }
+
+        @Override
+        public void accept(Double priority, WeightedSample sample) {
+            double newWeight = sample.weight * scalingFactor;
+            if (Double.compare(newWeight, 0) == 0) {
+                return;
+            }
+            WeightedSample newSample = new WeightedSample(sample.value, newWeight);
+            if (values.put(priority * scalingFactor, newSample) == null) {
+                count++;
+            }
+        }
+    }
+
+    private LockFreeExponentiallyDecayingReservoir(int size, double alpha, Duration rescaleThreshold, Clock clock) {
+        // Scale alpha to nanoseconds
+        this.alphaNanos = alpha * SECONDS_PER_NANO;
+        this.size = size;
+        this.clock = clock;
+        this.rescaleThresholdNanos = rescaleThreshold.toNanos();
+        this.state = new State(clock.getTick(), new AtomicLong(), new ConcurrentSkipListMap<>());
+    }
+
+    @Override
+    public int size() {
+        return (int) Math.min(size, state.count.get());
+    }
+
+    @Override
+    public void update(long value) {
+        long now = clock.getTick();
+        rescaleIfNeeded(now).update(value, now);
+    }
+
+    private State rescaleIfNeeded(long currentTick) {
+        State stateSnapshot = this.state;
+        long lastScaleTick = stateSnapshot.startTick;
+        if (currentTick - lastScaleTick >= rescaleThresholdNanos) {
+            State newState = stateSnapshot.rescale(currentTick);
+            if (stateUpdater.compareAndSet(this, stateSnapshot, newState)) {
+                // newState successfully installed
+                return newState;
+            }
+            // Otherwise another thread has won the race and we can return the result of a volatile read.
+            // It's possible this has taken so long that another update is required, however that's unlikely
+            // and no worse than the standard race between a rescale and update.
+            return this.state;
+        }
+        return stateSnapshot;
+    }
```

Comment:
minor JIT nit: might want to pull the `rescale` out to a separate method for readability and to keep hot path bytecode small
 
```suggestion
            return rescale(currentTick, stateSnapshot);
        }
        return stateSnapshot;
    }

    private State rescale(long currentTick, State stateSnapshot) {
        State newState = stateSnapshot.rescale(currentTick);
        if (stateUpdater.compareAndSet(this, stateSnapshot, newState)) {
            // newState successfully installed
            return newState;
        }
        // Otherwise another thread has won the race and we can return the result of a volatile read.
        // It's possible this has taken so long that another update is required, however that's unlikely
        // and no worse than the standard race between a rescale and update.
        return this.state;
    }
```

Reply:
addressed by https://github.com/dropwizard/metrics/pull/1656/commits/10f4c02ed82d59db3e414e4d5f0be4fb6ef65ef6

Reply:
Good catch, I've included before and after bytecode in the commit message: https://github.com/dropwizard/metrics/pull/1656/commits/10f4c02ed82d59db3e414e4d5f0be4fb6ef65ef6